File: test/cenote/test_water.py
from cenote.water import *
from cenote import config
import bungee

import unittest
from pint.testsuite import helpers


UREG = config.UREG


class TestPressureAtDepth(unittest.TestCase):
    """https://bluerobotics.com/learn/pressure-depth-calculator/"""

    def test_almost(self):
        VALUES = {
            bungee.Water.FRESH: {
                100 * UREG.meter: 141.81 * UREG.psi,
                32.81 * UREG.ft: 14.18 * UREG.psi,
            },
            bungee.Water.SALT: {
                100 * UREG.meter: 145.59 * UREG.psi,
                32.81 * UREG.ft: 14.56 * UREG.psi,
            },
        }
        PRESSURE_TOLERANCE = (1e-3 * UREG.atm).to(UREG.psi).magnitude

        for water in VALUES:
            for depth in VALUES[water]:
                expected_water_pressure = VALUES[water][depth]
                expected_pressure = expected_water_pressure + 1 * UREG.atm
                water_pressure = water_pressure_from_depth(depth, water)
                pressure = pressure_from_depth(depth, water)
                helpers.assert_quantity_almost_equal(
                    water_pressure, expected_water_pressure, PRESSURE_TOLERANCE
                )
                helpers.assert_quantity_almost_equal(
                    pressure, expected_pressure, PRESSURE_TOLERANCE
                )

File: backend/test/test_Water.cpp
#include <bungee/Constants.h>
#include <bungee/Water.h>
#include <gtest/gtest.h>
#include <map>

using namespace bungee;
using namespace units::literals;

namespace {

template <typename Unit> void EXPECT_UNIT_NEAR(Unit unit1, Unit unit2, Unit unitTol) {
    const double val1 = units::unit_cast<double>(unit1);
    const double val2 = units::unit_cast<double>(unit2);
    const double tol = units::unit_cast<double>(unitTol);
    EXPECT_NEAR(val1, val2, tol);
}

} // namespace

TEST(Water, Pressure) {
    static constexpr units::pressure::bar_t TOL(1e-3);
    // https://bluerobotics.com/learn/pressure-depth-calculator/
    const std::map<Water, std::map<units::length::meter_t, units::pressure::bar_t>> VALUES{
        {
            Water::FRESH,
            {
                {100_m, 9.777_bar},
            },
        },
        {
            Water::SALT,
            {
                {100_m, 10.038_bar},
            },
        }};
    for (auto const& [water, vals] : VALUES) {
        for (auto const& [depth, expectedWaterPressure] : vals) {
            const auto waterPressure = WaterPressureFromDepth(depth, water);
            const auto pressure = PressureFromDepth(depth, water);
            EXPECT_UNIT_NEAR(waterPressure, expectedWaterPressure, TOL);
            EXPECT_UNIT_NEAR(pressure, waterPressure + SURFACE_PRESSURE, TOL);
        }
    }
}